esp/services/ws_workunits/ws_wudetails.cpp

```diff
@@ -0,0 +1,349 @@
+
+        StatisticScopeType sst = queryScopeType(scopeType);
+        if (sst==SSTnone ||sst==SSTall)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid ScopeType(%s) in filterNestedScopeTypes",scopeType);
+
+        unsigned short usst = static_cast<unsigned short> (sst);
+        processedScopeTypeList.bAdd(usst, funcCompareUnsignedShort, isNew);
+    }
+}
+
+void CWUDetails::processWUStatsIntoResponse(IArrayOf<IEspWUScope> & respScopes,
+                                            IConstWUStatisticIterator & statsIter,
+                                            UnsignedShortArray & nestedScopeTypes,
+                                            UnsignedShortArray & returnAttribList,
+                                            StringArray & ids,
+                                            IArrayOf<CAttributeFilter> & arrayAttributeFilter,
+                                            unsigned __int64 minVersion,
+                                            unsigned __int64 & maxTimestamp)
+{
+    unsigned maxDepth = nfilter.getDepth();
+    Owned<CAttributeFilter> tmpForAttribSearch = new CAttributeFilter;
+    bool limitToIdList = ids.ordinality()>0;
+    bool returnSpecifyReturnAttribs = returnAttribList.ordinality()>0;
+    bool filterNestedScopeTypes = nestedScopeTypes.ordinality()>0;
+    bool filterByAttributeValue = arrayAttributeFilter.ordinality();
+
+    DBGLOG("Id list count: %d", ids.ordinality());
+    DBGLOG("ReturnAttribute  list count: %d", returnAttribList.ordinality());
+    DBGLOG("filterNestedScopeTypes count: %d", nestedScopeTypes.ordinality());
+    DBGLOG("AttributeFilter count: %d", arrayAttributeFilter.ordinality());
+
+    ids.sortAscii();      // Will need to lookup up 'id' in ids (for bSearch)
+    IArrayOf<IEspWUResponseAttribute> attribs;
+    SCMStringBuffer currentScope,currentScopeType,currentId;
+    unsigned statsCount = 0, statsAdded = 0;
+    ForEach(statsIter)
+    {
+        ++statsCount;
+        IConstWUStatistic & wuStat = statsIter.query();
+
+        SCMStringBuffer wuScope,  wuId;
+        wuStat.getScope(wuScope);
+        getScopeIdFromScopeName(wuId, wuScope);
+
+        if (limitToIdList && ids.bSearch(wuId.str(),funcCompareString)==NotFound)
+              continue;
+
+        if (maxDepth && getScopeDepth(wuScope) > maxDepth)
+            continue;
+
+        if (filterNestedScopeTypes && nestedScopeTypes.bSearch(wuStat.getScopeType(),funcCompareUnsignedShort)==NotFound)
+              continue;
+
+        if (filterByAttributeValue)
+        {
+            tmpForAttribSearch->setStatisticKind(wuStat.getKind());
+            aindex_t pos = arrayAttributeFilter.bSearch(tmpForAttribSearch,funcCompareAttribFilter);
+            if (pos!=NotFound)
+            {
+                CAttributeFilter & attributeFilter = arrayAttributeFilter.item(pos);
+                unsigned __int64 graphRawValue = wuStat.getValue();
+
+                if (attributeFilter.hasExactValue())
```

Comment:
encapsulation:  The large number of attributeFilter.xxx in the following code indicate the code is in the wrong place.  Better would be a function attributeFilter.matchesValue(graphRawValue);